File: sdk/go/provider/mailforspam.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const mailforspamBase = "https://api.mailforspam.com/api"
const mailforspamDefaultDomain = "mailforspam.com"

var mailforspamDomains = []string{"mailforspam.com", "tempmail.io", "disposable.email"}

func mailforspamHeaders(req *http.Request) {
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Referer", "https://mailforspam.com/")
	req.Header.Set("Origin", "https://mailforspam.com")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0")
}

func mailforspamRandomLocal() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 19)
	copy(b, "sdk")
	for i := 3; i < len(b); i++ {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

func mailforspamPickDomain(preferred *string) string {
	if preferred != nil {
		p := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(*preferred)), "@")
		for _, d := range mailforspamDomains {
			if strings.ToLower(d) == p {
				return d
			}
		}
	}
	return mailforspamDefaultDomain
}

func mailforspamString(v any) string {
	if v == nil {
		return ""
	}
	switch x := v.(type) {
	case string:
		return strings.TrimSpace(x)
	default:
		return strings.TrimSpace(fmt.Sprint(x))
	}
}

func mailforspamEmailsURL(email string, pageSize int) string {
	return fmt.Sprintf(
		"%s/mailboxes/%s/emails?page=1&page_size=%d&sort_by=date&sort_order=desc",
		mailforspamBase,
		url.PathEscape(email),
		pageSize,
	)
}

type mailforspamListResponse struct {
	Emails []map[string]any `json:"emails"`
}

func mailforspamGetJSON(u string) ([]byte, error) {
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	mailforspamHeaders(req)
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("mailforspam: http %d", resp.StatusCode)
	}
	return body, nil
}

func MailforspamGenerate(domain *string, channel ...string) (*CreatedMailbox, error) {
	ch := "mailforspam"
	if len(channel) > 0 && channel[0] != "" {
		ch = channel[0]
	}
	email := fmt.Sprintf("%s@%s", mailforspamRandomLocal(), mailforspamPickDomain(domain))
	if _, err := mailforspamGetJSON(mailforspamEmailsURL(email, 1)); err != nil {
		return nil, fmt.Errorf("mailforspam validate mailbox: %w", err)
	}
	return &CreatedMailbox{Channel: ch, Email: email}, nil
}

func mailforspamFetchMessage(id, email string) (map[string]any, error) {
	u := fmt.Sprintf(
		"%s/mailboxes/%s/emails/%s",
		mailforspamBase,
		url.PathEscape(email),
		url.PathEscape(id),
	)
	body, err := mailforspamGetJSON(u)
	if err != nil {
		return nil, err
	}
	var raw map[string]any
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, err
	}
	return raw, nil
}

func mailforspamFlatten(raw map[string]any, recipient string) map[string]any {
	to := mailforspamString(raw["receiver"])
	if to == "" {
		to = recipient
	}
	return map[string]any{
		"id":          raw["id"],
		"from":        mailforspamString(raw["sender"]),
		"to":          to,
		"subject":     raw["subject"],
		"text":        raw["body_text"],
		"html":        raw["body_html"],
		"date":        raw["date"],
		"isRead":      raw["readAt"] != nil,
		"attachments": raw["attachments"],
	}
}

func MailforspamGetEmails(email string) ([]NormEmail, error) {
	email = strings.TrimSpace(email)
	if email == "" {
		return nil, fmt.Errorf("mailforspam: empty email")
	}
	body, err := mailforspamGetJSON(mailforspamEmailsURL(email, 100))
	if err != nil {
		return nil, err
	}
	var data mailforspamListResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Emails))
	for _, item := range data.Emails {
		id := mailforspamString(item["id"])
		if id == "" {
			continue
		}
		detail, err := mailforspamFetchMessage(id, email)
		if err != nil {
			out = append(out, NormalizeMap(mailforspamFlatten(item, email), email))
			continue
		}
		out = append(out, NormalizeMap(mailforspamFlatten(detail, email), email))
	}
	return out, nil
}

File: sdk/npm/src/providers/mailforspam.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "mailforspam";
const API_BASE = "https://api.mailforspam.com/api";
const DEFAULT_DOMAIN = "mailforspam.com";
const DOMAINS = ["mailforspam.com", "tempmail.io", "disposable.email"];

const HEADERS: Record<string, string> = {
  Accept: "application/json",
  Referer: "https://mailforspam.com/",
  Origin: "https://mailforspam.com",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
};

interface MailforspamListResponse {
  emails?: MailforspamListMessage[];
}

interface MailforspamListMessage {
  id?: string;
  sender?: string;
  receiver?: string;
  subject?: string;
  date?: string;
  readAt?: string | null;
  attachments?: MailforspamAttachment[];
}

interface MailforspamDetail extends MailforspamListMessage {
  body_text?: string;
  body_html?: string;
}

interface MailforspamAttachment {
  filename?: string;
  size?: number;
  content_type?: string;
}

function mailboxEmailsUrl(email: string, pageSize = 100): string {
  return `${API_BASE}/mailboxes/${encodeURIComponent(email)}/emails?page=1&page_size=${pageSize}&sort_by=date&sort_order=desc`;
}

function randomLocal(): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let out = "sdk";
  for (let i = 0; i < 16; i++)
    out += chars[Math.floor(Math.random() * chars.length)];
  return out;
}

function pickDomain(preferred?: string | null): string {
  const p = (preferred || "").trim().replace(/^@/, "").toLowerCase();
  if (p) {
    const hit = DOMAINS.find((d) => d.toLowerCase() === p);
    if (hit) return hit;
  }
  return DEFAULT_DOMAIN;
}

function mapAttachments(
  raw?: MailforspamAttachment[],
): Record<string, unknown>[] {
  return Array.isArray(raw)
    ? raw.map((item) => ({
        filename: item.filename || "",
        size: item.size,
        contentType: item.content_type,
      }))
    : [];
}

function flattenDetail(
  raw: MailforspamDetail,
  recipient: string,
): Record<string, unknown> {
  return {
    id: raw.id || "",
    from: raw.sender || "",
    to: raw.receiver || recipient,
    subject: raw.subject || "",
    text: raw.body_text || "",
    html: raw.body_html || "",
    date: raw.date || "",
    isRead: Boolean(raw.readAt),
    attachments: mapAttachments(raw.attachments),
  };
}

function flattenList(
  raw: MailforspamListMessage,
  recipient: string,
): Record<string, unknown> {
  return {
    id: raw.id || "",
    from: raw.sender || "",
    to: raw.receiver || recipient,
    subject: raw.subject || "",
    date: raw.date || "",
    isRead: Boolean(raw.readAt),
    attachments: mapAttachments(raw.attachments),
  };
}

async function fetchMessage(
  id: string,
  email: string,
): Promise<MailforspamDetail | null> {
  const url = `${API_BASE}/mailboxes/${encodeURIComponent(email)}/emails/${encodeURIComponent(id)}`;
  const response = await fetchWithTimeout(url, {
    method: "GET",
    headers: HEADERS,
  });
  if (!response.ok) return null;
  const data = (await response.json()) as MailforspamDetail;
  return data && typeof data === "object" ? data : null;
}

export async function generateEmail(
  domain?: string | null,
  channel: Channel = CHANNEL,
): Promise<InternalEmailInfo> {
  const selectedDomain = pickDomain(domain);
  const email = `${randomLocal()}@${selectedDomain}`;
  const response = await fetchWithTimeout(mailboxEmailsUrl(email, 1), {
    method: "GET",
    headers: HEADERS,
  });
  if (!response.ok) {
    throw new Error(
      `mailforspam: validate mailbox failed HTTP ${response.status}`,
    );
  }

  return { channel, email };
}

export async function getEmails(email: string): Promise<Email[]> {
  const address = (email || "").trim();
  if (!address) throw new Error("mailforspam: empty email");

  const response = await fetchWithTimeout(mailboxEmailsUrl(address, 100), {
    method: "GET",
    headers: HEADERS,
  });
  if (!response.ok) {
    throw new Error(`mailforspam: list mailbox failed HTTP ${response.status}`);
  }

  const data = (await response.json()) as MailforspamListResponse;
  const list = Array.isArray(data.emails) ? data.emails : [];
  const emails: Email[] = [];
  for (const item of list) {
    if (!item.id) continue;
    const detail = await fetchMessage(String(item.id), address);
    emails.push(
      normalizeEmail(
        detail ? flattenDetail(detail, address) : flattenList(item, address),
        address,
      ),
    );
  }
  return emails;
}
